File: backend/internal/dto/review.go
package dto

import "time"

// ReviewRequest represents the payload to create a review
type ReviewRequest struct {
	Rating  int    `json:"rating" binding:"required,min=1,max=5"`
	Comment string `json:"comment" binding:"max=500"`
}

// ReviewResponse represents a published review returned to clients
type ReviewResponse struct {
	ID            string    `json:"id"`
	Rating        int       `json:"rating"`
	Comment       string    `json:"comment"`
	AuthorDisplay string    `json:"author_display"`
	CreatedAt     time.Time `json:"created_at"`
}

// ReviewListResponse is a paginated list of reviews
type ReviewListResponse struct {
	Items []ReviewResponse `json:"items"`
	Total int              `json:"total"`
	Page  int              `json:"page"`
	Limit int              `json:"limit"`
}

// ReviewSummary aggregates ratings for a product
type ReviewSummary struct {
	Average      float64     `json:"average"`
	Count        int         `json:"count"`
	Distribution map[int]int `json:"distribution"`
}

File: frontend/src/services/review.ts
import api from "./api";

export interface ReviewRequest {
  rating: number;
  comment: string;
}

export interface Review {
  id: string;
  rating: number;
  comment: string;
  author_id: string;
  author_display: string;
  created_at: string;
}

export interface ReviewListResponse {
  items: Review[];
  total: number;
  page: number;
  limit: number;
}

export interface ReviewSummary {
  average: number;
  count: number;
  distribution: Record<number, number>;
}

export async function listReviews(
  productSlug: string,
  params?: { page?: number; limit?: number; signal?: AbortSignal }
): Promise<ReviewListResponse> {
  const { page = 1, limit = 10, signal } = params || {};
  const { data } = await api.get<ReviewListResponse>(
    `/products/${productSlug}/reviews`,
    { params: { page, limit }, signal }
  );
  return data;
}

export async function getSummary(
  productSlug: string,
  opts?: { signal?: AbortSignal }
): Promise<ReviewSummary> {
  const { signal } = opts || {};
  const { data } = await api.get<ReviewSummary>(
    `/products/${productSlug}/reviews/summary`,
    { signal }
  );
  return data;
}

export async function createReview(
  productSlug: string,
  payload: ReviewRequest
): Promise<Review> {
  const { data } = await api.post<Review>(
    `/products/${productSlug}/reviews`,
    payload
  );
  return data;
}

export async function deleteReview(
  productSlug: string,
  reviewId: string
): Promise<void> {
  await api.delete(`/reviews/${reviewId}`);
}
